src/room.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright 2017 Tunnler Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "tunnler/room.h"
+
+#include "tunnler/room_message_types.h"
+
+/// Maximum number of concurrent connections allowed to this room.
+static const uint32_t MaxConcurrentConnections = 10;
+
+void Room::Create(const std::string& name, const std::string& server_address, uint16_t server_port) {
+    RakNet::SocketDescriptor socket(server_port, server_address.c_str());
+
+    server = RakNet::RakPeerInterface::GetInstance();
+    // TODO(Subv): Allow specifying the maximum number of concurrent connections.
+    server->Startup(MaxConcurrentConnections, &socket, 1);
+    server->SetMaximumIncomingConnections(MaxConcurrentConnections);
+
+    state = State::Open;
+
+    // Start a network thread to Receive packets in a loop.
+    room_thread = std::make_unique<std::thread>(&Room::ServerLoop, this);
+}
+
+void Room::Destroy() {
+    state = State::Closed;
+
+    {
+        std::lock_guard<std::mutex> lock(server_mutex);
+
+        if (server)
+            RakNet::RakPeerInterface::DestroyInstance(server);
+        server = nullptr;
+    }
+
```

Comment:
We should assert that we closed successfully. I'm not sure if raknet destroying the instance returns anything / could fail. But it would be quite horrendous if a host closes the server, returns a "Closed" state but might still have some socket open.

If raknet is fail-safe in this regard this is fine.